src/tools/miri/src/helpers.rs

```diff
@@ -479,6 +480,16 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
                 self.ecx
             }
 
+            fn aggregate_field_order(memory_index: &IndexVec<FieldIdx, u32>, idx: usize) -> usize {
+                // We need to do an *inverse* lookup: find the field that has position `idx` in memory order.
+                for (src_field, &mem_pos) in memory_index.iter_enumerated() {
```

Comment:
This loop is unfortunate, it makes aggregate traversal n^2 in the number of fields. But at least this only affects SB (and I doubt it is a limiting factor there). I didn't find a way to avoid this without introducing overhead for the common case (where we don't care about the order). The only nice way I came up with required return-position `impl trait` in a trait.